harness/determined/common/experimental/checkpoint/_checkpoint.py

```diff
@@ -153,6 +147,8 @@ def download(self, path: Optional[str] = None, mode: DownloadMode = DownloadMode
             mode (DownloadMode, optional): Governs how a checkpoint is downloaded. Defaults to
                 ``AUTO``.
         """
+        self.reload()
+        assert self.state
```

Comment:
it's for mypy. since `state` is being initialized elsewhere, mypy isn't smart enough to figure out that it should logically be non-None at certain points in the code.

Reply:
Interesting...

Can mypy work it out if the assertion is in `reload`?

Reply:
nope. mypy doesn't seem to infer types from inner function calls.

Reply:
> In this specific case, I don't see anything in this method where self.state can't be None.
mypy complains in this case because below this we check if the state of the checkpoint allows deletion. but maybe i misunderstood what you're saying?

Reply:
No... you understood correctly. I just confused my own self somehow giving Python the autovivifying semantics that Perl has. Sorry. Thanks for being patient.